src/iterators.jl

```diff
@@ -0,0 +1,301 @@
+    ind[i] can be either and index of w (and so in [1, G.n]) or be -1, meaning no index of w is 
+    stored at the position.
+
+    Note that inds is part of the iterator's state inds only to prevent reallocation in each 
+    iteration.
+    =#
+    rank = Int(1)
+    v = collect(1:G.k) 
+    inds = fill(-1, 3) 
+    (inds, (v, rank, inds))
+end
+
+@inline function Base.iterate(G::SubsetGrayCode, state)
+    # Based on Algorithm 2.13 in kreher1999combinatorial
+    v, rank, inds = state
+
+    if rank == G.len 
+        return nothing 
+    end
+    rank += 1
+
+    @inbounds begin
+        inds[1] = -1
+        inds[2] = -1
+        inds[3] = -1
+
+        i = 1
+        while (i <= G.k) && (v[i] == i) 
+            i = i + 1
+        end
+        if Base.rem(G.k - i, 2) != 0 
+            if (i == 1)
+                v[1] != (v[1] - 1) && _update_indexs!(inds, v[1], v[1] - 1) 
+                v[1] = (v[1] - 1)
+            else 
+                v[i - 1] != i && _update_indexs!(inds,v[i - 1], i) 
+                v[i - 1] = i
+                if i > 2
+                    v[i - 2] != (i - 1) && _update_indexs!(inds,v[i - 2], i - 1) 
+                    v[i - 2] = (i - 1)
+                end
+            end
+        else 
+            if i == G.k
+                if ( G.n != v[i] )
+                    if (i > 1)
+                      v[i - 1] != v[i] && _update_indexs!(inds,v[i - 1], v[i]) 
+                      v[i - 1] = v[i]
+                    end
+                    v[i] != (v[i] + 1) && _update_indexs!(inds,v[i], v[i] + 1) 
+                    v[i] = (v[i] + 1)
+                else
+                    v[i] != i && _update_indexs!(inds, v[i], i) 
+                    v[i] = i
+                end
+            else
+                if ( v[i + 1] != (v[i] + 1) )
+                    if i > 1
+                        v[i - 1] != v[i] && _update_indexs!(inds, v[i - 1], v[i]) 
+                        v[i - 1] = v[i]
+                    end
+                    v[i] != (v[i] + 1) && _update_indexs!(inds, v[i], v[i] + 1) 
+                    v[i] = (v[i] + 1)
+                else
+                    v[i + 1] != v[i] && _update_indexs!(inds, v[i + 1], v[i]) 
+                    v[i + 1] = v[i]
+                    v[i] != i && _update_indexs!(inds, v[i], i) 
+                    v[i] = i
+                end
+            end
+        end
+    end
+    return (inds, (v, rank, inds))
+end
+
+@inline function rest(G::SubsetGrayCode, rank)
+  #TODO there will be a future PR for integrating this (its used for multithreaded weight calculation)
+  #=
+  _subset_unrank(rank, G.n, vec)
+  inds = Vector{Int}([-1,-1,-1]) 
+  state = (vec, rank, inds)
+  return Base.rest(G, state)
+  =#
+end 
+
+function _update_indexs!(indexs::Vector{Int}, x::Int, y::Int)
+    _update_indexs!(indexs, x)
+    _update_indexs!(indexs, y)
+    return nothing
+end
+
+function _update_indexs!(indexs::Vector{Int}, x::Int)
+    if x == indexs[1]
```

Comment:
the update we're doing here corresponds to binary addition at the index ```x```. If ```x``` is already set in ```indexs``` this corresponds to two binary additions. These cancel out so we remove ```x``` from ```indexs```